File: Machine Learning Template Folder/Part 2 - Regression/Polynomial Regression/polynomial_regression.py

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

# Splitting the dataset into the Training set and Test set
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

#Fitting Linear Regression to dataset 

from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X,y)

#Predicting the result

y_pred = lin_reg.predict([6.5]) 

#Fitting polynominal Regression to the dataset

from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree=13)
X_poly = poly_reg.fit_transform(X)
lin_reg2 = LinearRegression()
lin_reg2.fit(X_poly,y)

#Predecting the result

y_pred2= lin_reg2.predict(poly_reg.fit_transform(6.5))

#Visualing the results
#Linear Model
plt.scatter(X, y, color = 'red')
plt.plot(X, lin_reg.predict(X), color = 'blue')
plt.title('Truth or bluff (Linear Regression) ')
plt.xlabel('Position Label')
plt.ylabel('Salary')
plt.show()
#Polynomial Model 

X_grid = np.arange(min(X),max(X),0.1)
X_grid = X_grid.reshape((len(X_grid),1))

plt.scatter(X, y, color = 'red') #red points, data sets ...
plt.plot(X_grid, lin_reg2.predict(poly_reg.fit_transform(X_grid)), color = 'blue') #Data predicted
plt.title('Truth or bluff (Polynomial Regression) ')
plt.xlabel('Position Label')
plt.ylabel('Salary')
plt.show()

File: Machine Learning Template Folder/Part 2 - Regression/Polynomial Regression/polynomial_regression.R
# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

#Fitting Linear Model

lin_reg = lm(formula = Salary ~ .,
             dataset)

#Fitting Polynomial Model

dataset$Level2 = dataset$Level ^ 2
dataset$Level3 = dataset$Level ^ 3
dataset$Level4 = dataset$Level ^ 4

poly_reg = lm(formula = Salary ~ .,
              data = dataset)

#Visualing Linear Regression results
library(ggplot2)
ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
            colour = 'blue') + 
  ggtitle("Truth or Bluff (Linear Regression)") +
  xlab("Level") + 
  ylab("Salary")

#Visualising Polynomila Regression results 

ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = dataset)),
            colour = 'blue') + 
  ggtitle("Truth or Bluff (Polynomial Regression)") +
  xlab("Level") + 
  ylab("Salary")


#Predecting results using Linear Model
data_test  = data.frame(Level = 6.5)
y_pred = predict(lin_reg, data_test)

#Predecting results using Polynomial Model

data_test  = data.frame(Level = 6.5,
                        Level2 = 6.5^2,
                        Level3 = 6.5^3,
                        Level4 = 6.5^4)
y_pred = predict(poly_reg, data_test)




